src/assets/TokenListController.ts

```diff
@@ -0,0 +1,288 @@
+
+  /**
+   * Creates a TokenListController instance
+   *
+   * @param options - Constructor options
+   * @param options.interval - The polling interval, in milliseconds
+   * @param options.messenger - A reference to the messaging system
+   * @param options.state - Initial state to set on this controller
+   */
+  constructor({
+    chainId,
+    onNetworkStateChange,
+    interval = DEFAULT_INTERVAL,
+    cacheRefreshThreshold = DEFAULT_THRESHOLD,
+    messenger,
+    state,
+  }: {
+    chainId: string;
+    onNetworkStateChange: (
+      listener: (networkState: NetworkState) => void,
+    ) => void;
+    interval?: number;
+    cacheRefreshThreshold?: number;
+    messenger: RestrictedControllerMessenger<
+      typeof name,
+      GetTokenListState,
+      TokenListStateChange,
+      never,
+      never
+    >;
+    state?: Partial<TokenListState>;
+  }) {
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.intervalDelay = interval;
+    this.cacheRefreshThreshold = cacheRefreshThreshold;
+    this.chainId = chainId;
+    onNetworkStateChange(async (networkState) => {
+      this.chainId = networkState.provider.chainId;
+      await safelyExecute(() => this.fetchTokenList());
+    });
+  }
+
+  /**
+   * Start polling for the token list
+   */
+  async start() {
+    await this.startPolling();
+  }
+
+  /**
+   * Stop polling for the token list
+   */
+  stop() {
+    this.stopPolling();
+  }
+
+  /**
+   * Prepare to discard this controller.
+   *
+   * This stops any active polling.
+   */
+  destroy() {
+    super.destroy();
+    this.stopPolling();
+  }
+
+  private stopPolling() {
+    if (this.intervalId) {
+      clearInterval(this.intervalId);
+    }
+  }
+
+  /**
+   * Starts a new polling interval
+   */
+  private async startPolling(): Promise<void> {
+    await safelyExecute(() => this.fetchTokenList());
+    this.intervalId = setInterval(async () => {
+      await safelyExecute(() => this.fetchTokenList());
+    }, this.intervalDelay);
+  }
+
+  /**
+   * Fetching token list from the Token Service API
+   */
+  async fetchTokenList(): Promise<void> {
+    const releaseLock = await this.mutex.acquire();
+    try {
+      const tokensFromAPI: Token[] = await safelyExecute(() =>
+        this.fetchFromCache(),
+      );
+      const { tokensChainsCache } = this.state;
+      const tokenList: TokenMap = {};
+
+      // filtering out tokens with less than 2 occurences
+      const filteredTokenList = tokensFromAPI.filter(
+        (token) => token.occurrences >= 2,
+      );
+      // removing the tokens with symbol conflicts
+      const symbolsList = filteredTokenList.map((token) => token.symbol);
+      const duplicateSymbols = [
+        ...new Set(
+          symbolsList.filter(
+            (symbol, index) => symbolsList.indexOf(symbol) !== index,
+          ),
+        ),
+      ];
+      const uniqueTokenList = filteredTokenList.filter(
```

Comment:
I'm wondering if you could simplify this section of code by using lodash's uniqBy?

```js
const uniqueTokenList = uniqBy(filteredTokenList, 'symbol');
```

This would select the first item by the 'symbol' key and discard subsequent finds in the same list. 

I could also be misinterpreting the end goal of tracking duplication so this is more of a NIT then a REQ for changes.

Reply:
Oh I see you're comparing duplicateSymbols with the token's name, versus its symbol. I'm not sure I understand that bit if you could elaborate on the goal here? In light of this, I think my previous comment is invalid.

Reply:
As part of MVP, we are removing all the tokens with symbols rather than keeping one and removing the rest. So I am filtering out all the symbols that have duplicates and then using that list I am removing all tokens from tokenlist with the corresponding symbols. 
I couldn't find a simpler way of doing it as this seems to be an unusual use-case, if there is any way to simplify this that ideas are most welcome.